torchtune/modules/transformer.py

```diff
@@ -113,50 +125,83 @@ def __init__(
         tok_embeddings: nn.Embedding,
         layer: TransformerDecoderLayer,
         num_layers: int,
+        max_seq_len: int,
+        num_heads: int,
+        head_dim: int,
         norm: nn.Module,
         output: nn.Linear,
     ) -> None:
         super().__init__()
+
         self.tok_embeddings = tok_embeddings
         self.layers = _get_clones(layer, num_layers)
         self.norm = norm
         self.output = output
+        self.max_seq_len = max_seq_len
+        self.num_heads = num_heads
+        self.head_dim = head_dim
+        self.causal_mask = None
 
-    def forward(
-        self, tokens: Tensor, mask: Optional[Tensor] = None, curr_pos: int = 0
-    ) -> Tensor:
+    def setup_caches(self, max_batch_size: int, dtype: torch.dtype) -> None:
```

Comment:
I'm curious about the choice to define this as a method on the `TransformerDecoder` class (especially with the addition of mandatory `max_seq_len`, `num_heads`, `head_dim` which are only needed for inference). Why not just define a utility method, e.g. 

```
def setup_caches(
	model: TransformerDecoder,
	max_batch_size: int,
	max_seq_len: int,
	num_heads: int,
	head_dim: int,
):
	# same body as setup_caches but with self -> model
```

Then optionally call this on the builder gated behind a flag. The slight drawback is that we do add one extra param to builders so it's not a 1:1 swap on the config side as it is now (maybe there's a way to do that though?). But honestly I think that's OK? Like if we are changing the model arch a bit for inference (which we are) there's no harm in being explicit about that.

I think the benefit is that anyone using `TransformerDecoder` directly (which should be a decent percentage of the people adding new models) does not have to worry about these extra params that are really only relevant for inference.

Reply:
This is a really interesting point. 

I think you touch upon this a bit. The reason I think having a method in the class makes sense is because the behavior of forward changes depending on whether you're calling it during training or during inference and I'd rather the model be able to handle this based on its internal state (eg: causal_mask). For example, if this is a utility, then you also need to take care of passing around the causal_mask explicitly in every call when this can be easily handled by the class itself? I also think, that having the forward of the decoder just take the input and the corresponding positions is nicer than explicitly passing the mask (though that might be a bit of a personal preference).

I'm actually not too concerned about num_heads, head_dim and max_seq_len because the model already has this information as part of all of its components - it's related to the model isn't it? In fact I think this is strictly better than passing something like max_batch_size which was only used during inference :) Let me know if that makes sense.